nbodykit/base/catalog.py

```diff
@@ -35,6 +216,36 @@ def __new__(cls, catalog, daskarray):
         self.attrs = {}
         return self
 
+    def __getitem__(self, key):
+        """
+        If ``key`` is an array-like index with the same length as the
+        underlying catalog, the slice will be performed with the
+        optimization provided by :func:`optimized_selection`.
+
+        Otherwise, the default behavior is returned.
+        """
+        d = None
+
+        # try to optimize the selection
+        sel = key[0] if isinstance(key, tuple) else key
+        if isinstance(sel, (list, numpy.ndarray, da.Array)):
+            if len(sel) == self.catalog.size:
+                if isinstance(sel, da.Array):
+                    sel = self.catalog.compute(sel)
+                try:
+                    d = optimized_selection(self, sel)
```

Comment:
This try except looks strange -- could you catch a DaskGraphOptimizationFailure instead?

Reply:
yeah that is a clearer solution